File: apps/desktop/src/m3u/group.ts
import { LucideFolder, LucideIcon, LucideTv } from "lucide-react";
import { ViewObject } from "./view-object";
import { TvShowWatchableObject, WatchableObject } from "./watchable";
import { M3UObject } from "./m3u";

export class GroupObject extends ViewObject {

  public Watchables: WatchableObject[];
  public Groups: GroupObject[];
  public CoverImages: CoverItem[] | undefined = undefined;
  public GetListIcon: LucideIcon = LucideFolder;

  constructor(name = "Group", ListIcon = LucideFolder) {
    super();
    this.Name = name;
    this.GetListIcon = ListIcon;
    this.Groups = [];
    this.Watchables = [];
  }

  /**
   * Clears the Watchables and Groups arrays.
   *
   * @param {} - No parameters.
   * @return {} - No return value.
   */
  public clear() {
    this.Watchables = [];
    this.Groups = [];
  }

  /**
   * Determines whether the given view object should be filtered.
   *
   * @param {ViewObject} v - The view object to be checked.
   * @return {boolean} Returns true if the view object should be filtered, false otherwise.
   */
  public shouldFilter(v: ViewObject): boolean {
    return this.UpperLevel?.shouldFilter(v) ?? false;
  }

  /**
   * Adds the given item to the list of watchable objects.
   *
   * @param {WatchableObject} item - The item to be added.
   */
  public justAdd(item: WatchableObject) {
    this.Watchables.push(item);
  }

  /**
   * Removes the specified item from the Watchables array.
   *
   * @param {WatchableObject} item - The item to be removed.
   */
  public justRemove(item: WatchableObject) {
    const index = this.Watchables.indexOf(item);
    if (index !== -1) {
      this.Watchables.splice(index, 1);
    }
  }

  /**
   * Returns the total count of local items.
   *
   * @return {number} The total count of local items.
   */
  public localCount(): number {
    return this.Groups.length + this.Watchables.length;
  }

  /**
   * Performs the last check on all groups and sorts the groups and watchables.
   *
   * @return {void} 
   */
  public lastCheck() {
    for (let i = this.Groups.length - 1; i >= 0; i--) {
      const gr = this.Groups[i];
      gr.lastCheck();
      if (gr.TotalCount == 0)
        this.Groups.splice(i, 1);
    }
    const compare = (a: ViewObject, b: ViewObject) => {
      if (a.isSticky && !b.isSticky) return -1;
      if (!a.isSticky && b.isSticky) return 1;

      if (a.AddedDate && b.AddedDate) {
        const diff = b.AddedDate.getTime() - a.AddedDate.getTime();
        if (diff !== 0) return diff;
      } else if (a.AddedDate) {
        return -1;
      } else if (b.AddedDate) {
        return 1;
      }
      return a.Name.localeCompare(b.Name);
    };
    this.Groups.sort(compare);
    this.Watchables.sort(compare);
  }

  /**
   * Retrieves a list of cover images.
   *
   * @param {number} [limit=9] - The maximum number of cover images to retrieve.
   * @return {CoverItem[]} - An array of cover items representing the cover images.
   */
  public getImageList(limit: number = 9): CoverItem[] {
    if (this.CoverImages) return this.CoverImages;
    const result: CoverItem[] = [];
    const generateNumbers = (max: number, n: number) => {
      const result: number[] = [];
      const set: number[] = [];
      for (let i = 0; i < max; i++) set.push(i);
      if (n >= max) return set;   //return whole set
      for (let i = 0; i < n; i++) {
        const rand = Math.floor(Math.random() * set.length);
        const num = set.splice(rand, 1)[0];
        result.push(num);
      }
      return result;
    }

    const list = generateNumbers(this.Watchables.length, limit);
    for (const i of list) {
      const w = this.Watchables[i];
      result.push(new CoverItem(w.Name, w.Logo, w.isHot));
    }

    if (result.length < limit) {
      const temp: CoverItem[] = [];
      for (const g of this.Groups) temp.push(...g.getImageList());
      const list = generateNumbers(temp.length, limit - result.length);
      for (const i of list)
        result.push(temp[i]);
    }

    this.CoverImages = result;
    return this.CoverImages;
  }

  get TotalCount(): number {
    return this.Groups.reduce((total, g) => total + g.TotalCount, 0) +
      this.Watchables.length;
  }

  get LiveStreamCount(): number {
    return this.Groups.reduce((total, g) => total + g.LiveStreamCount, 0) +
      this.Watchables.reduce((total, w) => w.category === 'LiveStream' ? total + 1 : total, 0);
  }

  get TvShowSeasonCount(): number {
    let total: number = 0;
    for (const m of this.Groups) {
      if (m instanceof TvShowGroupObject) {
        total += m.seasonCount;
      } else if (m instanceof GroupObject) {
        total += m.TvShowSeasonCount;
      }
    }
    return total;
  }

  get TvShowEpisodeCount(): number {
    let total = 0;
    for (const m of this.Groups) {
      if (m instanceof TvShowGroupObject) {
        total += m.episodeCount;
      } else if (m instanceof GroupObject) {
        total += m.TvShowEpisodeCount;
      }
    }
    return total;
  }

  get MovieCount(): number {
    let total = 0;
    for (const m of this.Groups) {
      if (m instanceof TvShowGroupObject ||
        m instanceof TvShowSeasonGroupObject)
        continue;
      if (m instanceof GroupObject)
        total += (<GroupObject>m).MovieCount;
    }
    return this.Watchables.reduce((total, w) => w.category === 'Movie' ? total + 1 : total, total);
  }

  get TvShowCount(): number {
    return this.Groups.reduce((total, g) => g instanceof TvShowGroupObject ? total + 1 : total + g.TvShowCount, 0);
  }

  private setFrom(watchable: WatchableObject, m3u: M3UObject) {
    watchable.Url = m3u.url;
    watchable.Logo = m3u.logo ?? "";
    watchable.Name = m3u.title;
    watchable.Year = m3u.year;

    if (watchable instanceof TvShowWatchableObject) {
      watchable.Season = m3u.season ?? 1;
      watchable.Episode = m3u.episode ?? 1;
    }
  }

  public hasItem(item: WatchableObject): boolean {
    const group = this.Groups.find(g => g.Title === item.Title);
    return !!(group && group.Watchables.some(w => w.Url === item.Url));
  }

  public has(item: M3UObject): boolean {
    const group = this.Groups.find(g => g.Title === item.title);
    return !!(group && group.Watchables.some(w => w.Url === item.url));
  }

  public Add(m3u: M3UObject): WatchableObject {
    const obj = new WatchableObject();
    this.setFrom(obj, m3u);
    obj.UpperLevel = this;
    return this.AddWatchable(obj);
  }

  public AddWatchable(watchable: WatchableObject): WatchableObject {
    this.Watchables.push(watchable);
    return watchable;
  }

  public findByUrl(url: string): WatchableObject | undefined {
    for (const g of this.Groups) {
      const obj = g.findByUrl(url);
      if (obj) return obj;
    }
    return this.Watchables.find(w => w.Url === url) as WatchableObject;
  }

  public AddTvShow(m3u: M3UObject): WatchableObject {
    const tvShowGroup = this.addTvGroup(m3u.title);
    const tvShowSeazon = tvShowGroup.addSeason(m3u.season ?? 1);
    const tvShow = new TvShowWatchableObject();
    this.setFrom(tvShow, m3u);
    tvShow.UpperLevel = tvShowSeazon;
    tvShowSeazon.Watchables.push(tvShow);
    return tvShow;
  }

  addGroup(groupName: string): GroupObject {
    if (!groupName) {
      groupName = "unnamed group";
    }

    const group = this.Groups.find(g => g.Name === groupName);

    if (group) {
      return group;
    }

    const groupObject = new GroupObject();
    groupObject.Name = groupName;
    groupObject.UpperLevel = this;
    this.Groups.push(this.modifyGroupBeforeAdded(groupObject));
    return groupObject;
  }

  addTvGroup(groupName: string): TvShowGroupObject {
    if (!groupName) {
      groupName = "unnamed tvshow";
    }

    const group = this.Groups.find(g => g.Name === groupName);

    if (group) {
      return <TvShowGroupObject>group;
    }

    const groupObject = new TvShowGroupObject();
    groupObject.Name = groupName;
    groupObject.UpperLevel = this;
    this.Groups.push(this.modifyGroupBeforeAdded(groupObject));
    return groupObject;
  }

  public modifyGroupBeforeAdded(group: GroupObject): GroupObject {
    return group;
  }

  toString(): string {
    let str: string = "";
    for (const g of this.Groups) {
      str += g.Name + " " + g.Groups.length + " " + g.Watchables.length + "</br>";
    }
    return str;
  }

  public searchProgress(group: Ref<GroupObject>, parts: string[], abortController: AbortController) {
    for (const g of this.Groups) {
      if (abortController.signal.aborted) return;
      if (g instanceof TvShowGroupObject) {
        if (exString.hasMatch(parts, g.Name))
          group.value.Groups.push(g);
      }
      else
        g.searchProgress(group, parts, abortController);
    }
    for (const w of this.Watchables) {
      if (abortController.signal.aborted) return;
      if (exString.hasMatch(parts, w.Name))
        group.value.Watchables.push(w);
    }
  }

}

export class TvShowGroupObject extends GroupObject {

  get episodeCount(): number {
    return this.Groups.reduce((last, m) => m instanceof TvShowSeasonGroupObject ?
      last + (<TvShowSeasonGroupObject>m).episodeCount : last, this.Watchables.length);
  }

  get seasonCount(): number {
    return this.Groups.length;
  }

  addSeason(season: number): TvShowSeasonGroupObject {
    season = Math.max(1, season);
    const name = `Season ${season}`;

    const group = this.Groups.find(g => g.Name === name);

    if (group) {
      return group as TvShowSeasonGroupObject;
    }

    const groupObject = new TvShowSeasonGroupObject();
    groupObject.Season = season;
    groupObject.Name = name;
    groupObject.UpperLevel = this;
    this.Groups.push(groupObject);
    return groupObject;
  }

  getSeason(season: number): TvShowSeasonGroupObject | undefined {
    return this.Groups.find(g => g instanceof TvShowSeasonGroupObject && g.Season === season) as TvShowSeasonGroupObject;
  }

  getEpisode(season: number, episode: number): WatchableObject | undefined {
    return this.getSeason(season)?.getEpisode(episode);
  }

}

export class TvShowSeasonGroupObject extends GroupObject {
  constructor() {
    super();
    this.GetListIcon = LucideTv;
  }

  public getEpisode(episode: number): WatchableObject | undefined {
    return this.Watchables.find(w =>
      w instanceof TvShowWatchableObject &&
      (<TvShowWatchableObject>w).Episode === episode
    )
  }

  public Season: number = 0;
  get episodeCount(): number { return this.Watchables.length; }
}


class CoverItem {
  public Logo: string;
  public Name: string;
  public isHot: boolean;

  constructor(name: string, logo: string, isHot: boolean) {
    this.Name = name;
    this.isHot = isHot;
    this.Logo = logo;
  }
}

File: apps/mobile/lib/models/group.dart
import 'package:flutter/material.dart';

import 'view_object.dart';
import 'watchable.dart';
import 'm3u_object.dart';
import '../utils/string_matcher.dart';

/// Cover image item for group thumbnails.
/// Mirrors: shared/content/src/models/group.ts → CoverItem
class CoverItem {
  final String name;
  final String logo;
  final bool isHot;

  const CoverItem(this.name, this.logo, {this.isHot = false});
}

/// A group that contains watchables and/or sub-groups.
/// Mirrors: shared/content/src/models/group.ts → GroupObject
class GroupObject extends ViewObject {
  List<WatchableObject> watchables = [];
  List<GroupObject> groups = [];
  List<CoverItem>? coverImages;

  GroupObject([String groupName = 'Group']) {
    name = groupName;
  }

  /// Mirrors: GroupObject.GetListIcon in desktop
  IconData get listIcon => Icons.folder_outlined;

  void clear() {
    watchables = [];
    groups = [];
  }

  bool shouldFilter(ViewObject v) => upperLevel?.shouldFilter(v) ?? false;

  void justAdd(WatchableObject item) => watchables.add(item);

  void justRemove(WatchableObject item) => watchables.remove(item);

  void removeWatchable(WatchableObject watchable) => justRemove(watchable);

  int localCount() => groups.length + watchables.length;

  /// Removes empty sub-groups and sorts everything.
  /// Mirrors: GroupObject.lastCheck()
  void lastCheck() {
    for (var i = groups.length - 1; i >= 0; i--) {
      groups[i].lastCheck();
      if (groups[i].totalCount == 0) groups.removeAt(i);
    }
    int compare(ViewObject a, ViewObject b) {
      if (a.isSticky && !b.isSticky) return -1;
      if (!a.isSticky && b.isSticky) return 1;
      if (a.addedDate != null && b.addedDate != null) {
        final diff = b.addedDate!.compareTo(a.addedDate!);
        if (diff != 0) return diff;
      } else if (a.addedDate != null) {
        return -1;
      } else if (b.addedDate != null) {
        return 1;
      }
      return a.name.compareTo(b.name);
    }

    groups.sort(compare);
    watchables.sort(compare);
  }

  /// Random cover images from watchables (up to [limit]).
  List<CoverItem> getImageList([int limit = 9]) {
    if (coverImages != null) return coverImages!;
    final result = <CoverItem>[];

    final indices = _randomIndices(watchables.length, limit);
    for (final i in indices) {
      final w = watchables[i];
      result.add(CoverItem(w.name, w.logo, isHot: w.isHot));
    }

    if (result.length < limit) {
      final temp = <CoverItem>[];
      for (final g in groups) { temp.addAll(g.getImageList()); }
      final extra = _randomIndices(temp.length, limit - result.length);
      for (final i in extra) { result.add(temp[i]); }
    }

    coverImages = result;
    return result;
  }

  int get totalCount =>
      groups.fold(0, (sum, g) => sum + g.totalCount) + watchables.length;

  int get liveStreamCount =>
      groups.fold(0, (sum, g) => sum + g.liveStreamCount) +
      watchables.where((w) => w.category == M3UCategory.liveStream).length;

  int get tvShowSeasonCount {
    var total = 0;
    for (final g in groups) {
      if (g is TvShowGroupObject) {
        total += g.seasonCount;
      } else {
        total += g.tvShowSeasonCount;
      }
    }

    return total;
  }

  int get tvShowEpisodeCount {
    var total = 0;
    for (final g in groups) {
      if (g is TvShowGroupObject) {
        total += g.episodeCount;
      } else {
        total += g.tvShowEpisodeCount;
      }
    }
    return total;
  }

  int get movieCount {
    var total = 0;
    for (final g in groups) {
      if (g is TvShowGroupObject || g is TvShowSeasonGroupObject) continue;
      total += g.movieCount;
    }
    return total +
        watchables.where((w) => w.category == M3UCategory.movie).length;
  }

  int get tvShowCount => groups.fold(
        0,
        (sum, g) => g is TvShowGroupObject ? sum + 1 : sum + g.tvShowCount,
      );

  bool hasItem(WatchableObject item) {
    final g = groups.where((g) => g.title == item.title).firstOrNull;
    return g != null && g.watchables.any((w) => w.url == item.url);
  }

  bool has(M3UObject item) {
    final g = groups.where((g) => g.title == item.title).firstOrNull;
    return g != null && g.watchables.any((w) => w.url == item.url);
  }

  WatchableObject add(M3UObject m3u) {
    final obj = WatchableObject();
    _setFrom(obj, m3u);
    obj.upperLevel = this;
    return addWatchable(obj);
  }

  WatchableObject addWatchable(WatchableObject watchable) {
    watchables.add(watchable);
    return watchable;
  }

  WatchableObject? findByUrl(String url) {
    for (final g in groups) {
      final obj = g.findByUrl(url);
      if (obj != null) return obj;
    }
    return watchables.where((w) => w.url == url).firstOrNull;
  }

  WatchableObject addTvShow(M3UObject m3u) {
    final tvShowGroup = addTvGroup(m3u.title);
    final tvShowSeason = tvShowGroup.addSeason(m3u.season ?? 1);
    final tvShow = TvShowWatchableObject();
    _setFrom(tvShow, m3u);
    tvShow.upperLevel = tvShowSeason;
    tvShowSeason.watchables.add(tvShow);
    return tvShow;
  }

  GroupObject addGroup(String groupName) {
    if (groupName.isEmpty) groupName = 'unnamed group';
    final existing = groups.where((g) => g.name == groupName).firstOrNull;
    if (existing != null) return existing;
    final g = GroupObject(groupName);
    g.upperLevel = this;
    groups.add(modifyGroupBeforeAdded(g));
    return g;
  }

  TvShowGroupObject addTvGroup(String groupName) {
    if (groupName.isEmpty) groupName = 'unnamed tvshow';
    final existing = groups.where((g) => g.name == groupName).firstOrNull;
    if (existing != null) return existing as TvShowGroupObject;
    final g = TvShowGroupObject(groupName);
    g.upperLevel = this;
    groups.add(modifyGroupBeforeAdded(g));
    return g;
  }

  GroupObject modifyGroupBeforeAdded(GroupObject g) => g;

  void searchProgress(
      GroupObject result, List<String> parts, bool Function() isAborted) {
    for (final g in groups) {
      if (isAborted()) return;
      if (g is TvShowGroupObject) {
        if (StringMatcher.hasMatch(parts, g.name)) result.groups.add(g);
      } else {
        g.searchProgress(result, parts, isAborted);
      }
    }
    for (final w in watchables) {
      if (isAborted()) return;
      if (StringMatcher.hasMatch(parts, w.name)) result.watchables.add(w);
    }
  }

  void _setFrom(WatchableObject watchable, M3UObject m3u) {
    watchable.url = m3u.url;
    watchable.logo = m3u.logo ?? '';
    watchable.name = m3u.title;
    watchable.year = m3u.year;
    watchable.category = m3u.category;
    if (watchable is TvShowWatchableObject) {
      watchable.season = m3u.season ?? 1;
      watchable.episode = m3u.episode ?? 1;
    }
  }

  static List<int> _randomIndices(int max, int n) {
    if (max == 0) return [];
    final pool = List.generate(max, (i) => i);
    pool.shuffle();
    return pool.take(n).toList();
  }
}

/// A TV show (contains seasons as sub-groups).
/// Mirrors: shared/content/src/models/group.ts → TvShowGroupObject
class TvShowGroupObject extends GroupObject {
  TvShowGroupObject([super.groupName = 'TvShow']);

  @override
  String get title => 'Tv Shows';

  @override
  IconData get listIcon => Icons.tv;

  int get episodeCount => groups.fold(
        watchables.length,
        (sum, g) =>
            g is TvShowSeasonGroupObject ? sum + g.episodeCount : sum,
      );

  int get seasonCount => groups.length;

  TvShowSeasonGroupObject addSeason(int season) {
    season = season < 1 ? 1 : season;
    final seasonName = 'Season $season';
    final existing =
        groups.where((g) => g.name == seasonName).firstOrNull;
    if (existing != null) return existing as TvShowSeasonGroupObject;
    final g = TvShowSeasonGroupObject();
    g.season = season;
    g.name = seasonName;
    g.upperLevel = this;
    groups.add(g);
    return g;
  }

  TvShowSeasonGroupObject? getSeason(int season) =>
      groups
          .whereType<TvShowSeasonGroupObject>()
          .where((g) => g.season == season)
          .firstOrNull;

  TvShowWatchableObject? getEpisode(int season, int episode) =>
      getSeason(season)?.getEpisode(episode);
}

/// A season within a TV show (contains episodes as watchables).
/// Mirrors: shared/content/src/models/group.ts → TvShowSeasonGroupObject
class TvShowSeasonGroupObject extends GroupObject {
  int season = 0;

  @override
  IconData get listIcon => Icons.calendar_view_week_outlined;

  int get episodeCount => watchables.length;

  TvShowWatchableObject? getEpisode(int episode) => watchables
      .whereType<TvShowWatchableObject>()
      .where((w) => w.episode == episode)
      .firstOrNull;
}
